R/helper.R

```diff
@@ -98,19 +96,58 @@ run_app <- function(app_name, qc_dir, lib_path, config_path) {
     url <- sprintf("http://127.0.0.1:%s", port)
 
 
+    new_rstudioapi <- {
+      version_diff <- compareVersion(packageDescription("rstudioapi", fields = "Version"), "0.16-0")
+      if (version_diff < 0) {
+        FALSE
+      }
+      else {
+        TRUE
+      }
+    }
+
+
+    sp1 <- cli::make_spinner()
     cli::cli_inform("Waiting for shiny app to start...")
+
+    total_time <- 35  # Total wait time in seconds
+    interval <- 0.1   # Spinner refresh interval in seconds
+    iterations <- total_time / interval  # Total iterations needed
     counter <- 1
-    while(counter < 20) {
+    while(counter <= iterations) {
+      sp1$spin()
       if (is_shiny_ready(url)) {
-        cli::cli_inform("Shiny app started")
+        sp1$finish()
+        cli_alert_success("Shiny app started")
         break
       }
 
+      # if rstudioapi is >= 0.16.0, can check if there's been an error in the bgj
+      if (new_rstudioapi) {
+        if (rstudioapi::jobGetState(job_id) == "failed") {
+          sp1$finish()
+          cli::cli_alert_danger("Shiny app could not be started due to error (see Background Jobs panel)")
+          break
+        }
+      }
+
+
       counter <- counter + 1
-      Sys.sleep(1)
+      Sys.sleep(interval)
     }
-    if (counter > 19) cli::cli_alert_danger("Shiny app could not be started due to timeout or error")
 
+    # check if there was a timeout
+    if (counter > iterations) {
+      # if rstudio api is newer, it caught it eariler in the case of an error, so it's definitely a timeout
+      if (new_rstudioapi) {
+        sp1$finish()
```

Comment:
being very picky, but you can move the sp1$finish in here and 147 outside of the if-else